### PR TITLE
MM-57013 Added download button for downloading logs from server logs page in system console

### PR DESCRIPTION
**Pull request by Aryakoste**
#### Summary

Added download logs button in server logs page of admin in system console. When clicked logs get downloaded as mattermost.log file.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/26320
Jira https://mattermost.atlassian.net/browse/MM-57013

#### Screenshots

|  before  |  after  |
|----|----|
| 
plainLogsBefore
 | 
plainLogsAfter
 |

|  before  |  after  |
|----|----|
| 
logsBefore
 | 
logsAfter
 |

#### Release Note

```release-note
Added download functionality for admins to download server logs from Server Logs Page in System Console.
```

### CODE REVIEW

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hanzei) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:184]: One thing to note here is that in plain view, the download button "only" downloads the current page of logs. On the formated JSON page, the download button seams to download everything.

Can we bring both modes into feature parity or at least make it clear in the UI which logs are getting downloaded?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for the contribution @Aryakoste :+1: 

The code looks fine, but I have concerns about the UX. Please see my comment below.

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (Aryakoste) [reply webapp/channels/src/components/admin_console/server_logs/logs.tsx:184]: So should I make it download current page logs for both or download all logs for both plain and JSON formatted.

Comment (svelle) [reply webapp/channels/src/components/admin_console/server_logs/logs.tsx:184]: @hanzei I agree, I think all logs should be downloaded.

Thanks for the PR @Aryakoste !

Comment (Aryakoste) [reply webapp/channels/src/components/admin_console/server_logs/logs.tsx:189]: Done.

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hanzei) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:102]: When testing the feature, I noticed that the download when using the plain view as a leading empty line. Can that get removed?

File: webapp/platform/client/src/client4.ts
Comment (hanzei) [inline webapp/platform/client/src/client4.ts:3112]: I'm surprised that no pagination is needed, but this call seems to fetch the whole log file.

**Review by hanzei — APPROVED** [review]

**Comment by hanzei** [comment]
@svelle Would you mind taking a look from an UX standpoint?

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hanzei) [reply webapp/channels/src/components/admin_console/server_logs/logs.tsx:102]: Hm, let me re-test tomorrow

**Review by svelle — APPROVED** [review]
Looks good to me from a UX standpoint!

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hanzei) [reply webapp/channels/src/components/admin_console/server_logs/logs.tsx:102]: The first image is from the plain logs.

Screenshot from 2024-03-19 08-57-42

Second image is from the JSON logs
Screenshot from 2024-03-19 08-58-19

Please also note the different orderings in the first image.

Both issues (different ordering, leading empty line) are not a huge deal. But it would be nice to have them fixed.

**Review by hanzei — APPROVED** [review]
There are some non-blocking issues with the change, still LGTM :+1:

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hmhealey) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:97]: Instead of going to all the work of downloading and stringifying the logs here, what if we were to actually add a `download=true` parameter to the `getLogs` API and then have the download button trigger that API with the download parameter set along with whatever the page normally calls to load the logs? To do that, you have to set some headers on the API response, but there's a `web.WriteFileResponse` helper function on the server to handle all that for you.

The APIs for file attachments and some other things behave similarly, and it feels like it would be a more proper way of doing this since we're relying mostly on standard HTTP headers and browser behaviour to process the logs instead of doing it in JS which seems like it could cause problems when the logs are larger on a busier server. The server could even potentially stream the results, but that feels like that might require more work which is outside the scope of this.

Thoughts on all this? CCing @hanzei as well since this would be a pretty big change in the approach

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hmhealey) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:101]: Assuming we don't move this to the server though, instead of storing all of the plain logs in the Redux store and passing them in as a prop, could we actually change this to something more like this where we just get the response directly from the action? I don't think we need to save all the logs in the Redux state since this is a one-off action, so we don't need to store them in Redux state
```js
            const response = await this.props.actions.getAllPlainLogs();
            if (!response.data) {
                // handle error somehow
            }
            logString = response.data.join('');
```

I'm a bit surprised that this worked as-is. I know we await on the action, but I would've expected `this.props.allPlainLogs` to have been the logs from when `downloadLogs` is first called which would be before we actually loaded them

File: webapp/channels/src/packages/mattermost-redux/src/actions/admin.ts
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/actions/admin.ts:67]: Again, assuming we don't move more of this to the server, I'd rather not store this data in redux state since it's only used once. If you remove the `onSuccess` argument from here, it won't save anything in Redux, but you can still get the result by from the return value from the action which would be an `ActionResult<string[]>`

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (Aryakoste) [reply webapp/channels/src/components/admin_console/server_logs/logs.tsx:97]: Yeah good point. This is an appropriate way tbh. As when logs will be larger it will be heavy to process on frontend. I can make this change. If you are raising a new issue related to this, you can assign it to me directly. Or I can do in this issue only.

Comment (hanzei) [reply webapp/channels/src/components/admin_console/server_logs/logs.tsx:97]: 2/5 that adding a separate `/logs/download` endpoint makes sense as the other `/logs/` endpoint do some kind of parsing on the logs.

**Comment by Aryakoste** [comment]
Should I make backend change as discussed or do the frontend changes given.

**Comment by hmhealey** [comment]
If you could do the backend change, that'd be great

**Comment by hmhealey** [comment]
Yeah, a new API for each of those would work. The body of the new methods will be mostly the same as the old ones except for the headers attached to the response to trigger the download. The API metrods can call [`web.WriteFileResponse`](https://github.com/mattermost/mattermost/blob/master/server/platform/shared/web/files.go#L37) with true for its `forceDownload` parameter to write all those headers for you.

**Comment by Aryakoste** [comment]
Update:- I have made the api and checked the response.  Just have to make some changes on frontend to make it properly work. Will submit PR after that.

**Comment by Aryakoste** [comment]
Update:- I have made the api and checked the response. Just have to make some changes on frontend to make it properly work. Will commit my changes after that.

**Comment by Aryakoste** [comment]
func downloadLogs(c *Context, w http.ResponseWriter, r *http.Request) {

	lines, _ := c.App.GetLogs(c.AppContext, c.Params.Page, c.Params.LogsPerPage)

	tmpfile, err := os.CreateTemp("", "logs-*.log")
	if err != nil {
		http.Error(w, "Failed to create temporary file", http.StatusInternalServerError)
		return
	}
	defer tmpfile.Close()
	defer os.Remove(tmpfile.Name())

	for _, line := range lines {
		if _, err := io.WriteString(tmpfile, line); err != nil {
			http.Error(w, "Failed to write logs to file", http.StatusInternalServerError)
			return
		}
	}

	fileInfo, err := tmpfile.Stat()
	if err != nil {
		http.Error(w, "Failed to get file information", http.StatusInternalServerError)
		return
	}

	web.WriteFileResponse("mattermost.log",
		"text/plain",
		fileInfo.Size(),
		time.Now(),
		*c.App.Config().ServiceSettings.WebserverMode,
		tmpfile,
		true,
		w,
		r)
}

this is the function/controller i wrote to download the log file. But the log file isnt downloading automatically. I have set the forceDownload parameter to true.

**Comment by hmhealey** [comment]
That's strange. I would've expected that to work, but perhaps something else is missing.

Could you push those changes so that I could try them out? They can either be part of this branch or on a separate one, but I want to be able to run the code locally

**Comment by Aryakoste** [comment]
I have pushed the changes. Check and let me know.

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hmhealey) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:165]: The reason that this isn't loading the logs now is that, instead of a button that runs JS to fetch the file, you want this to be an anchor element with the href being the API you added.
```
<a
    className='btn btn-primary'
    href={Client4.getUrl() + 'api/v4/logs/download'}
    target='_blank'
>
</a>
```
When you click on that and the browser tries to navigate to it, it'll instead open a new tab and download the log file. At least on Firefox, that worked for me with the server code you had

File: server/channels/api4/system.go
Comment (hmhealey) [inline server/channels/api4/system.go:410]: Instead of using a temp file for this, you can put all of this into a `ByteReader` to avoid writing to and then reading from disk. `ByteReader` also implements the `io.ReadCloser` interface that `WriteFileResponse` needs. I think that'll look like this:

```go
	lines, _ := c.App.GetLogs(c.AppContext, c.Params.Page, c.Params.LogsPerPage)
	b := bytes.NewReader([]byte(model.ArrayToJSON(lines)))

	web.WriteFileResponse("mattermost.log",
		"text/plain",
		int64(b.Len()),
		time.Now(),
		*c.App.Config().ServiceSettings.WebserverMode,
		b,
		true,
		w,
		r)
```

Comment (hmhealey) [reply server/channels/api4/system.go:410]: Alternatively, you could probably open the actual log file and serve that. I tried that, and while it was much quicker on my local instance, it also tried downloading my entire 70MB log file, and I don't know if we want to do that

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hmhealey) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:19]: Could you fix these linting errors? I'm pretty sure that most of these can be fixed automatically if you run `npm run fix`

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hmhealey) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:161]: For this one, I guess we have to use the `ExternalLink` component for these buttons. It renders a regular `a` element, and it automatically adds `target="_blank"` and the `rel` when needed

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/admin.ts
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/reducers/entities/admin.ts:35]: Could you remove the logic for `allPlainLogs` and `getAllPlainLogs` now that that's no longer needed?

**Review by hmhealey — CHANGES_REQUESTED** [review]
This works great. Thanks for making that change. There's a few linting issues remaining and some now-unused code that can be cleaned up, but this is just about ready to go

**Review by hmhealey — APPROVED** [review]
Thanks for fixing all those. Assuming CI passes, this looks good to me!

**Comment by hmhealey** [comment]
@hanzei I'm not sure who your team usually assigns to QA test your tickets, so I assigned @yasserfaraazkhan. If there's someone else who'd be better to test it, feel free to re-assign

**Comment by yasserfaraazkhan** [comment]
@Aryakoste can you check the CI failure ?

**Comment by Aryakoste** [comment]
what does this check-i18n suggest me to do.

**Comment by hmhealey** [comment]
Looks like that that's complaining about the order of the translation strings being wrong. I can update that

**Comment by hmhealey** [comment]
Whoops. It turns out that there were also server-side translation strings that were out of order as well. I'll make sure CI passes this time instead of just hoping it does 😅

**Review by yasserfaraazkhan — APPROVED** [review]

**Review by hanzei — DISMISSED** [review]
Sorry for the late review @Aryakoste! I did review the changes earlier before you added the backend stuff. Hence, I want to chime in with another review.

**Comment by Aryakoste** [comment]
I have documented the API. Writing the test but wanted to ask in test i should check response status, content header along with the response content with mock logs right. And wanted suggestion on how could i get the logs directly from cluster info or from file as you suggested. I am having trouble finding that. I have found some functions but they take pagination params into consideration. Do you have any suggestions on this.  @hanzei @hmhealey

**Comment by hmhealey** [comment]
@hanzei's currently out on vacation, but I'll see if I can answer those while we find someone else to finish off the second review.

- For the tests, that sounds good to me. The only other thing I could think to test would maybe be the permissions to ensure that only admins can access the API. You can take a look at the tests for the `GetFile` API for a reference for how those tests might look too
- I'll ask around to see where we should get those logs from. We could potentially stream the log file directly to the caller of the API, but I don't know if that's an option in a multi-server environment

File: server/channels/api4/system.go
Comment (pvev) [reply server/channels/api4/system.go:1]: +1 on this suggestion. For the test for the function, as @hmhealey suggested, testing the permissions and role restrictions handling make sense. Extra it sounds good to check the response status, maybe content headers, and maybe response content using mock logs.

an example on how to test this could be: https://github.com/mattermost/mattermost/blob/master/server/channels/app/support_packet_test.go#L341

**Review by pvev — CHANGES_REQUESTED** [review]
Thank you @Aryakoste for your contribution 🚀 . I have added some small requests to finish polishing your changes and some guidance for the testing and from where to get the log file. Please, let us know if you have any further questions.

File: server/channels/api4/system.go
Comment (hmhealey) [reply server/channels/api4/system.go:419]: Huh, I assumed it was something more complicated than just reading directly from the log file because I wouldn't expect that to work in a multi-server environment. I guess not.

In that case, take a look at the code linked by @pvev to find how to get the log file location, but instead of using `os.ReadFile` to read the file into memory as an array of bytes, you can probably use `os.Open` to get a reader which can be passed directly to `WriteFileResponse`

**Comment by Aryakoste** [comment]
@hmhealey @pvev @hanzei  All changes are done. Just the test is giving some issue. I have tried different methods of writing this test but it goes FAIL. Idk why i am having hard time writing this test XDD. Need your guyz suggestions.

**Comment by Aryakoste** [comment]
I have written the tests just need help fixing it.

**Review by hmhealey — CHANGES_REQUESTED** [review]

**Comment by hmhealey** [comment]
Hopefully that answers your questions about the tests, but let me know if it doesn't. I could also make make those changes for you if you want to help get that last bit finished up so this can be merged

**Comment by Aryakoste** [comment]
@hmhealey  Changes done. Thanks

File: api/v4/source/logs.yaml
Comment (hmhealey) [inline api/v4/source/logs.yaml:2]: The one CI check is failing because it the type for this is incorrect
```suggestion
  get:
```

File: webapp/channels/src/i18n/en.json
Comment (hmhealey) [inline webapp/channels/src/i18n/en.json:1465]: Another CI check is failing because this string isn't in the right place because this file is sorted alphabetically. If you go to the `webapp` folder and run `npm run i18n-extract -w channels`, it'll regenerate this file and sort it into the right order

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hmhealey) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:161]: The final CI failure is because these `ExternalLink` components are missing a `location` prop. I think you can just set that to `'download_logs'` to make the type checker happy

File: webapp/channels/src/components/admin_console/server_logs/logs.tsx
Comment (hmhealey) [inline webapp/channels/src/components/admin_console/server_logs/logs.tsx:198]: Same here

**Review by hmhealey — CHANGES_REQUESTED** [review]
The tests and all the code looks good to me now, but there's a few CI things to fix still

**Comment by Aryakoste** [comment]
Changes done @hmhealey

**Comment by hmhealey** [comment]
Thanks! Those changes look like they should fix CI, but I need to update the branch and fix the merge conflicts to confirm that

**Comment by hmhealey** [comment]
@Aryakoste I'm not sure why one of those checks is failing now. It's complaining that the new API isn't documented when it looks like it should be to me.

Do you mind if I make some changes to see if I can fix it?

**Comment by Aryakoste** [comment]
@hmhealey Yeah sure. But i have documented the API and made changes in Makefile as its a new yaml file. Then why the CI failure.

**Comment by hmhealey** [comment]
It looks like the issue was the indentation level of your docs. The script that generates the HTML for the API reference concatenates the yaml files together, and the file you added wasn't indented as much, so the resulting yaml file was invalid because of that. That should be fixed now

**Review by hmhealey — APPROVED** [review]

**Comment by Aryakoste** [comment]
Ok got it. Thanks

**Review by pvev — APPROVED** [review]
LGTM! thanks for all the efforts with this PR @Aryakoste 🚀 🎆

**Comment by hmhealey** [comment]
Thanks for all the work you did on this PR! At this point, it should be in for the 9.11 release in August

**Pull request merged**
